An administrative user can authenticate themselves to access an "admin" area of the service

robbiepaul opened the pull request:
So that DfE staff in the Financial Incentives Team will be able to access the data relating to claims and process them, we need an authentication mechanism that will grant them access to a secure part of the application.

This story is to do the initial plumbing that will allow a user to sign in using DfE Sign-in to access an "/admin" part of the application. For now, this will simply contain a holding page with a "sign-out" link that ends the session.

The identifier and secret are stored in Heroku. Place them in a `.env.local` file and run `bin/start` to start the dev server with https.

Note: Anyone with access to the 'Teacher Payments' service in DfE Sign In will be able to sign in, we currently don't check which organisation is returned. We may want to do this check in the future depending on how access to services is managed.

https://trello.com/c/A7865IBw

erbridge commented on config/environments/development.rb:
```diff
@@ -52,10 +52,26 @@
   config.assets.quiet = true
 
+  # Force https
+  config.force_ssl = true
```
Does this also enable HSTS on local machines? If so, it makes it hard to run other things on `localhost` that don't use HTTPS.

mec replied:
Looks like we can disable the HSTS part, by setting:

`config.ssl_options = { hsts: false }`

https://api.rubyonrails.org/v5.2.3/classes/ActionDispatch/SSL.html

@erbridge do you fancy exlpaining why HSTS could be an issue? I've never come accross it before?

erbridge replied:
HSTS basically tells your browser to always redirect that URL to HTTPS. This is fine in real environments, since once you add SSL, you rarely remove it, but for `localhost`, it means that your browser will try to upgrade all `http://localhost` requests to `https://localhost`, which might not be there. And it can be a real pain to get your browser to forget that once it's happened.

robbiepaul replied:
@erbridge Good catch, I've just checked and local https will still work without this config setting so I've removed it

erbridge replied:
We should be forcing SSL in production, though. How about something like:
```
config.force_ssl = true
config.ssl_options = { hsts: Rails.env.production? }
```

mec commented on docs/architecture-decisions/0007-use-dfe-sign-in-for-admin-authentication.md:
```diff
@@ -0,0 +1,26 @@
+DfE Sign In accounts can only be created by a small number of 'approvers' in
+DfE. 
+
+Developers will be required to run development with https as DfE Sing In 
```
Type here, 'Sing'

tekin commented on config/initializers/omniauth.rb:
```diff
@@ -0,0 +1,24 @@
+OmniAuth.config.logger = Rails.logger
+
+dfe_sign_in_issuer_uri    = URI(ENV.fetch("DFE_SIGN_IN_ISSUER"))
```
I guess calling `fetch` here means the app will fall over in environments where these variables are not defined, e.g. the current temporary prod environment on Heroku. Just noting that here as it means we'll have to add the variables to the prod environment before this code is promoted there.

tekin commented on bin/start:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
```
For the sake of following a more Rails-ish convention, maybe we should use a `Procfile` with foreman instead of a custom bash script for this. We can create a `Procfile.dev` and a [`.foreman`](https://ddollar.github.io/foreman/#DEFAULT-OPTIONS) file to point to this file as the default to use locally without confusing Heroku with local-development only stuff in a generic`Procfile`.

robbiepaul replied:
Thanks for this, I've learnt something new today 👍 I've amended the commit with the change

robbiepaul commented on README.md:
```diff
@@ -20,6 +20,24 @@ Architecture decision records can be found in the
 
+### DfE Sign In credentials
+
+By default in development OmniAuth will run in test mode. This means that you 
```
Ah, I have that setting on in Rubymine but it seems a bit inconsistent. I'll have a look if there's a way I can highlight trailing whitespace

robbiepaul replied:
It turns out the official Rubymine Markdown plugin was overriding this setting, after disabling the plugin it's now stripping the whitespace off the end of the line in markdown files 👍

robbiepaul commented:
I've added the refactor as a separate commit for now to make reviewing easier. If you're happy with it I'll squash it together with the other commit.

The main side effect of the change is that we'll need to request to get our callback URLs changed with DfE Sign In, which is fine I just thought I'd note it here.

erbridge commented on README.md:
```diff
@@ -16,9 +16,36 @@ Architecture decision records can be found in the
+     section.
+2. Run `bundle install` to install the gem dependencies
+3. Run `bundle exec rails db:setup` to set up the database development
+4. Run `foreman start` to launch the app on https://localhost:3000
```
Any chance we could binstub `foreman`?

tekin replied:
I've suggested we add it to the `Gemfile` `development` group so we can run it with bundler instead.

robbiepaul replied:
I've updated the readme to say `bundle exec foreman start`

robbiepaul commented:
I'm gonna hold off merging this PR until the callback URL has changed in DfE Sign In, I've requested it via email (30/5). I'll also add the missing environment vars to Heroku

robbiepaul commented:
I thought I would do one last test of this locally and it seems that `foreman` ignores the `.env.local` file, preferring `.env`. 

So I've made some changes to reflect that: 
- added `.env` to the gitignore and updated readme
- added `.env.development` (for running the app without DfE Sign In) and `.env.test` (for CI)

The redirect URL has been changed on DfE Sign In and I've added the environment vars to Heroku staging, so this is ready to go
